pyomo/network/port.py

```diff
@@ -0,0 +1,606 @@
+        timer.report()
+
+    def _initialize_members(self, initSet):
+        for idx in initSet:
+            tmp = self[idx]
+            for key in self._implicit:
+                tmp.add(None, key)
+            if self._extends:
+                for key, val in iteritems(self._extends.vars):
+                    tmp.add(val, key, self._extends._rules[key][0])
+            if self._initialize:
+                self._add_from_container(tmp, self._initialize)
+            if self._rule:
+                items = apply_indexed_rule(
+                    self, self._rule, self._parent(), idx)
+                self._add_from_container(tmp, items)
+
+    def _add_from_container(self, port, items):
+        if type(items) is dict:
+            for key, val in iteritems(items):
+                if type(val) is tuple:
+                    port.add(val[0], key, val[1])
+                else:
+                    port.add(val, key)
+        else:
+            for val in self._initialize:
+                if type(val) is tuple:
+                    port.add(val[0], rule=val[1])
+                else:
+                    port.add(val)
+
+    def _pprint(self, ostream=None, verbose=False):
+        """Print component information."""
+        def _line_generator(k, v):
+            for _k, _v in sorted(iteritems(v.vars)):
+                if _v is None:
+                    _len = '-'
+                elif hasattr(_v, '__len__'):
+                    _len = len(_v)
+                else:
+                    _len = 1
+                yield _k, _len, str(_v)
+        return (
+            [("Size", len(self)),
+             ("Index", self._index if self.is_indexed() else None)],
+             iteritems(self._data),
+             ( "Name", "Size", "Variable"),
+             _line_generator)
+
+    def display(self, prefix="", ostream=None):
+        """
+        Print component state information
+
+        This duplicates logic in Component.pprint()
+        """
+        if not self.active:
+            return
+        if ostream is None:
+            ostream = sys.stdout
+        tab = "    "
+        ostream.write(prefix + self.local_name + " : ")
+        ostream.write("Size=" + str(len(self)))
+
+        ostream.write("\n")
+        def _line_generator(k,v):
+            for _k, _v in sorted(iteritems(v.vars)):
+                if _v is None:
+                    _val = '-'
+                elif not hasattr(_v, 'is_indexed') or not _v.is_indexed():
+                    _val = str(value(_v))
+                else:
+                    _val = "{%s}" % (
+                        ', '.join('%r: %r' % (
+                            x, value(_v[x])) for x in sorted(_v._data)))
+                yield _k, _val
+        tabular_writer(ostream, prefix+tab,
+                       ((k, v) for k, v in iteritems(self._data)),
+                       ("Name", "Value"), _line_generator)
+
+    def Equality(port, name, index_set):
```

Comment:
I just read up on the difference between class and static methods [here](https://stackoverflow.com/questions/12179271/meaning-of-classmethod-and-staticmethod-for-beginner), and it looks like I would prefer static methods. Class methods require that the first argument be a reference to a class object, but that would be unnecessary here since I don't need a handle on `Port` as an argument. Not only that, it would change the api of passable rule functions to have to be something that takes a class as its first argument which is weird because it's unnecessary, and then that affects users who may want to define their own custom expansion function and use that as the rule.

That being said, I do think it would look cleaner to use the decorators instead of the `staticmethod` function calls below, so I'll probably switch to using that.